File: app/src/main/java/com/chess/cryptobot/market/Market.java
package com.chess.cryptobot.market;

import com.chess.cryptobot.exceptions.MarketException;
import com.chess.cryptobot.model.History;
import com.chess.cryptobot.model.response.CurrenciesResponse;
import com.chess.cryptobot.model.response.OrderBookResponse;
import com.chess.cryptobot.model.response.TickerResponse;
import com.chess.cryptobot.model.response.TradeLimitResponse;

import java.util.List;

public interface Market {
    String BITTREX_MARKET = "bittrex";
    String LIVECOIN_MARKET = "livecoin";

    String getMarketName();

    Double getAmount(String coinName) throws MarketException;

    OrderBookResponse getOrderBook(String pairName) throws MarketException;

    List<? extends TickerResponse> getTicker() throws MarketException;

    List<CurrenciesResponse> getCurrencies() throws MarketException;

    TradeLimitResponse getMinQuantity() throws MarketException;

    String getAddress(String coinName) throws MarketException;

    void sendCoins(String coinName, Double amount, String address) throws MarketException;

    void buy(String pairName, Double price, Double amount) throws MarketException;

    void sell(String pairName, Double price, Double amount) throws MarketException;

    boolean keysIsEmpty();

    List<History> getOpenOrders() throws MarketException;

    List<History> getHistory() throws MarketException;
}

File: app/src/main/java/com/chess/cryptobot/market/Market.kt
package com.chess.cryptobot.market

import android.content.Context
import com.chess.cryptobot.exceptions.MarketException
import com.chess.cryptobot.model.History
import com.chess.cryptobot.model.response.CurrenciesResponse
import com.chess.cryptobot.model.response.OrderBookResponse
import com.chess.cryptobot.model.response.TickerResponse
import com.chess.cryptobot.model.response.TradeLimitResponse

interface Market {
    fun getMarketName(): String

    fun resetBalance()

    @Throws(MarketException::class)
    fun getAmount(coinName: String): Double

    @Throws(MarketException::class)
    fun getOrderBook(pairName: String): OrderBookResponse

    @Throws(MarketException::class)
    fun getTicker(): List<TickerResponse>

    @Throws(MarketException::class)
    fun getCurrencies(): List<CurrenciesResponse>

    @Throws(MarketException::class)
    fun getMinQuantity(): TradeLimitResponse?

    @Throws(MarketException::class)
    fun getAddress(coinName: String): String?

    @Throws(MarketException::class)
    fun sendCoins(coinName: String, amount: Double, address: String)

    @Throws(MarketException::class)
    fun buy(pairName: String, price: Double, amount: Double)

    @Throws(MarketException::class)
    fun sell(pairName: String, price: Double, amount: Double)

    fun keysIsEmpty(): Boolean
    @Throws(MarketException::class)
    fun getOpenOrders(): List<History>

    @Throws(MarketException::class)
    fun getHistory(context: Context): List<History>

    @Throws(MarketException::class)
    fun getDepositHistory(): List<History>

    @Throws(MarketException::class)
    fun getWithdrawHistory(): List<History>

    companion object {
        const val POLONIEX_MARKET = "poloniex"
        const val BITTREX_MARKET = "bittrex"
        const val BINANCE_MARKET = "binance"
        const val BINANCE_WITHDRAWAL_MARKET = "binanceWithdrawal"

    }
}